### PR TITLE
feat(bundler): add `nsis`, closes #4450, closes #2319  

### LINKED ISSUE #2319: perUser InstallScope
**Is your feature request related to a problem? Please describe.**
I am unsure if there are technical reasons for this not being possible, but wish to raise it to see if it is a possibility.

**Describe the solution you'd like**
Could the windows installer support the perUser InstallScope (and I think maybe it'd need the limited InstallPrivileges) wix settings so that a Tauri app would not require admin privileges to install?

**Describe alternatives you've considered**
I appreciate WebView2 will have it's own limitations, but if that were already installed, could the app itself be less restricted in installation options?

**Additional context**
N/A

### LINKED ISSUE #4450: [feat] Add NSIS support
### Describe the problem

Wix toolset, a bit difficult to use, ugly interface, maybe I'm not using it properly. 

### Describe the solution you'd like

Can you provide more packaged installation tools? such as NSIS,Squirrel.Windows

### Alternatives considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by amrbashir**
### What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?

- [ ] Yes, and the changes were approved in issue #___
- [X] No

### Checklist
- [x] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [x] I have added a convincing reason for adding this feature, if necessary

### TODOs:

- [x] resources
- [x] external binaries
- [x] set AppUserModelID for notifications, jumplists, grouping...etc
- [x] downgrades and upgrades
- [x] Webview2 install modes
- [x] Prompt to kill if the app is running
- [x] updater package
- [x] change file
- [ ] deep link

### CODE REVIEW

**Comment by FabianLars** [comment]
The last thing i can think of would be support for other version formats than just `major.minor.patch`.

I just disabled the version check to test it, and it somewhat works as-is. Just noticed 2 problems (prob more i didn't find?):
1. the version in `programs & features` is wrapped in quotes `"version"`
2. Checking the exe properties: product version has the correct version (in this case 0.1.0-alpha) but file version doesn't (0.1.0.0 in this case) - not sure if these are system limitations

Do you know nsis (and potentially windows) limitations/specs here? i'm having difficulties reading the nsis docs, i didn't find any mentions about this.

**Comment by amrbashir** [comment]
> 1. the version in `programs & features` is wrapped in quotes `"version"`

That is actually my fault. It seems that I was wrapping the version in quotes in the registry key.

> 2. Checking the exe properties: product version has the correct version (in this case 0.1.0-alpha)  file version doesn't (0.1.0.0 in this case) - not sure if these are system limitations 

I think this is a Windows limitation.

I think we can allow using `major.minor.patch-pre` format but we would have to write a custom `VersionCompare` function because the built-in doesn't handle that format well, which is needed to detect if we should upgrade, downgrade or re-install

File: core/config-schema/schema.json
Comment (seahindeniz) [inline core/config-schema/schema.json]: Ah, just saw this!

If it's possible: what about having a third state, and making the value more understandable without a boolean value?
- `"allUsers"` - All users
- `"currentUser"` - Current user
- Accepts object
  - `"target"` : `"both" | "allUsers" | "currentUser"`
  - `"recommendedOption"` or `"highlightedOption"` : `"allUsers" | "currentUser"`

The third object will make both options visible and recommend an item, instead of forcing to install the app to a pre-selected target.

Comment (amrbashir) [reply core/config-schema/schema.json]: The thing is if your app logic depends on writing some files next to the exe, you can't use `all_users` so I wanted the system-wide option to be explicitly opted-in, otherwise your users might report unexpected errors because they installed it to a location that doesn't have write access. Also, that the installer will have to be launched as administrator even if the user plan to use `current_user`.

Comment (amrbashir) [reply core/config-schema/schema.json]: I think I might change this option to be

* `"all_users"` - All users
* `"current_user"` - Current user
* `"both"`

Although this might complicate the logic of the installer. Thanks for the feedback.

Comment (seahindeniz) [reply core/config-schema/schema.json]: The elevated access may also be needed and can be implemented, and I already stumbled upon an issue and after discussing with Fabian on Discord, this issue got created #5584.

> Also, that the installer will have to be launched as administrator even if the user plan to use current_user

This kind of concerns me. Some users may not have an admin privilege and still want to install the app. I'm assuming there's no way around that as well?

Comment (amrbashir) [reply core/config-schema/schema.json]: > > Also, that the installer will have to be launched as administrator even if the user plan to use current_user
> 
> This kind of concerns me. Some users may not have an admin privilege and still want to install the app. I'm assuming there's no way around that as well?

This will only be a concern if you use `all_users` or `both` option when building the installer. Unfortunately, I couldn't find a way around that.

Comment (amrbashir) [reply core/config-schema/schema.json]: The only way I found is to write 3 installers and combine, one that prompts users to choose `current_user` or `all_users` and includes the other two installers inside it and depending on that, it writes the appropriate installer to disk in `%TEMP%` and launches one of the other two installers. Which is messy and will pollute the installer size.

**Comment by lucasfernog** [comment]
You made it look easy @amrbashir let me know if I did something wrong. Let's merge this tomorrow :heart:

**Comment by amrbashir** [comment]
There are two features I was hoping to address before merge:
- [X] Removing the version limitation and allow `1.0.0-alpha.1`
- [ ] Add combined install modes

I will see if I can finish them tomorrow, if not, I can add them later in another PR.

**Comment by lucasfernog** [comment]
We can add it later too no problem.

**Comment by amrbashir** [comment]
Sure go ahead and merge if you want, I will add the missing features laeter.

File: tooling/bundler/src/bundle/windows/templates/SemverCompare.nsh
Comment (lucasfernog) [inline tooling/bundler/src/bundle/windows/templates/SemverCompare.nsh]: Any chance we can use a generic logic instead of only allowing alpha/beta as prerelease name? So instead of checking for `== "beta"` we check for `!= ""` and check if `$V4` is lower than `$VR4` (with StrCmp) instead of comparing with `"alpha"`.

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/SemverCompare.nsh]: yeah, I wasn't sure how semver treats other preleases other than `beta` and `alpha` so thought I would hard code them. I will try with your suggestions.

Comment (lucasfernog) [reply tooling/bundler/src/bundle/windows/templates/SemverCompare.nsh]: it's always a string comparison. `alpha < beta < rc` for instance.

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/SemverCompare.nsh]: alright, this is finally fixed. Feel free to merge as is. I will work on the other feature in another PR.

**Comment by amrbashir** [comment]
took me a lot of trials and errors and workarounds but `both` modes are finally supported. This PR is ready to be merged.

File: core/tauri-utils/src/config.rs
Comment (FabianLars) [inline core/tauri-utils/src/config.rs]:
```suggestion
  /// Default mode for the installer.
  ///
  /// Install the app by default in a directory that doesn't require Administrator access.
  ///
  /// Installer metadata will be saved under the `HKCU` registry path.
  CurrentUser,
  /// Install the app by default in the `Program Files` folder directory requires Administrator
  /// access for the installation.
  ///
  /// Installer metadata will be saved under the `HKLM` registry path.
  PerMachine,
  /// Combines both modes and allows the user to choose at install time
  /// whether to install for the current user or per machine. Note that this mode
  /// will require Administrator access even if the user wants to install it for the current user only.
  ///
  /// Installer metadata will be saved under the `HKLM` or `HKCU` registry path based on the user's choice.
```
Meant to fix only a few typos (`ve` - `be`) but adjusted the wording a little bit, but i can revert that if you want.
Oh and schema.json of course needs the same changes.

Comment (amrbashir) [reply core/tauri-utils/src/config.rs]: yeah I like that.

**Review by lucasfernog — APPROVED** [review]

**Pull request merged**
